### PR TITLE
Cross-dataset: gc=1.0+WD=1e-2 — AirfRANS regularization recipe transfer to DrivAerML

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS recently achieved a breakthrough (0.001236 surface MSE, -16.4% vs prior) using gradient clipping gc=1.0 and weight decay WD=1e-2 with AdamW. The DrivAerML baseline has never been tested with these regularization techniques. We hypothesize that the same stabilizing effects (reduced gradient blow-up at cosine LR peaks, better generalization via weight decay) will transfer to the 3D DrivAerML setting and push surface_rel_l2_pct below 4.619%. We also sweep cosine T_max and LR to find the best compound on DrivAerML.

GPU 2 is the primary transfer test: gc=1.0 + WD=1e-2 on the existing DrivAerML golden config. GPUs 3-7 explore compound variations on DrivAerML to isolate each lever. GPUs 0-1 re-confirm the AirfRANS and TandemFoil baselines with the same recipe for consistency.

## Instructions

Run all 8 GPUs in parallel from `target/icml2026/`. Use `--wandb-group gc-wd-transfer` on all runs. Do NOT hardcode `SENPAI_TIMEOUT_MINUTES` or `SENPAI_MAX_EPOCHS`.

**GPU 0 - TandemFoil baseline confirmation:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --model-slices 64 \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --wandb-group gc-wd-transfer
```

**GPU 1 - AirfRANS baseline confirmation:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --optimizer adamw --lr 7e-4 \
  --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --enable-fourier \
  --epochs 999 --wandb-group gc-wd-transfer
```

**GPU 2 - DrivAerML: gc=1.0 + WD=1e-2 transfer (PRIMARY TEST):**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier \
  --epochs 999 --wandb-group gc-wd-transfer
```

**GPU 3 - DrivAerML: gc=1.0 + WD=1e-2 + T_max=20:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 20 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier \
  --epochs 999 --wandb-group gc-wd-transfer
```

**GPU 4 - DrivAerML: gc=1.0 + WD=5e-3 (lighter decay):**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=4 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 5e-3 \
  --no-use-ema --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier \
  --epochs 999 --wandb-group gc-wd-transfer
```

**GPU 5 - DrivAerML: gc=1.0 + WD=1e-2 + lr=4e-4:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=5 python train.py \
  --dataset drivaerml --optimizer adamw --lr 4e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier \
  --epochs 999 --wandb-group gc-wd-transfer
```

**GPU 6 - DrivAerML: gc=1.0 + WD=1e-2 + lr=6e-4:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=6 python train.py \
  --dataset drivaerml --optimizer adamw --lr 6e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier \
  --epochs 999 --wandb-group gc-wd-transfer
```

**GPU 7 - DrivAerML: gc=1.0 only, no WD (isolate clipping effect):**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=7 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 \
  --no-use-ema --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier \
  --epochs 999 --wandb-group gc-wd-transfer
```

## Baseline

**DrivAerML (primary target):**
- `val_primary/surface_rel_l2_pct = 4.619%` (epoch 256, PR #2691)
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, no gc, no WD, 267 epochs, 180-min budget
- Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --model-layers 4 --model-hidden-dim 512 --model-heads 8 --enable-fourier --epochs 999`

**AirfRANS:**
- `val_primary/surface_mse = 0.001236` (epoch 349, PR #2828)
- Config: 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5

**TandemFoil:**
- `val_primary/surface_pressure_mae = 30.10` (epoch 157, PR #2810)
- Config: 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT brook:

## Results

### Cross-dataset regularization transfer: gc=1.0 + WD=1e-2

All 8 GPUs ran in parallel with `--wandb-group gc-wd-transfer`. SENPAI_TIMEOUT_MINUTES=360, SENPAI_MAX_EPOCHS=999 (not hardcoded — from env).

**Important caveat:** Running 6 concurrent DrivAerML processes caused significant I/O contention. DrivAerML runs completed only ~94-101 epochs vs the baseline's 267 epochs (at 180 min solo). The DrivAerML results below are at roughly 1/3 the baseline's training iterations.

---

### AirfRANS (GPU 1): NEW BEST — 0.001031 (-16.6%)

| Metric | Baseline (PR #2828) | This run | Change |
|--------|-------------------|----------|--------|
| val_primary/surface_mse | 0.001236 | **0.001031** | **-16.6%** |
| Best epoch | 349 (of 388) | 343 (of 352) | — |
| External target | 0.0043 | 0.001031 | **76.0% below** |

- Config: 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5, Fourier, no-EMA
- Extended training (360 min vs 180 min baseline) pushed deeper into the same optimization basin
- W&B: `6wv5py96`
- Command: `CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-layers 2 --model-hidden-dim 256 --model-heads 4 --enable-fourier --epochs 999 --wandb-group gc-wd-transfer --agent brook`
- Peak memory: ~50 GB (GPU 1)

### TandemFoil (GPU 0): No improvement — 31.31 (+4.0%)

| Metric | Baseline (PR #2810) | This run | Change |
|--------|-------------------|----------|--------|
| val_primary/surface_pressure_mae | 30.10 | 31.31 | +4.0% |
| Best epoch | 157 | 121 (of 137) | — |

- Config: 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10, Fourier+physics, no-EMA
- I/O contention reduced epoch count (137 vs 157+ in baseline). At epoch 121, still on a descending trajectory but the baseline had 20+ more epochs to find its best.
- W&B: `loxokopy`
- Peak memory: ~22 GB (GPU 0)

### DrivAerML (GPUs 2-7): Transfer FAILED — all variants miss baseline by 70%+

| GPU | Config | Best val % | Best ep | Total ep | W&B |
|-----|--------|-----------|---------|----------|-----|
| **6** | gc=1.0+WD=1e-2+**lr=6e-4** | **7.884%** | 93 | 94 | `rd6dskn9` |
| **2** | gc=1.0+WD=1e-2+T30 (PRIMARY) | **8.091%** | 100 | 101 | `b85peepu` |
| 7 | gc=1.0 only, no WD (isolate) | 8.111% | 90 | 94 | `hhtj1tax` |
| 3 | gc=1.0+WD=1e-2+T20 | 8.161% | 101 | 101 | `zhjzpxan` |
| 4 | gc=1.0+WD=**5e-3** | 8.274% | 95 | 97 | `kysy2p7g` |
| 5 | gc=1.0+WD=1e-2+**lr=4e-4** | 16.442% | 36 | 92 | `1xvy15zz` |

**Baseline: 4.619% at epoch 256 (PR #2691, 4L/512d, no gc, no WD, 267 epochs, 180 min)**

#### DrivAerML training dynamics — catastrophic divergence

The most important finding is that **gc=1.0 is completely insufficient for DrivAerML's gradient dynamics**:

- **GPU 2 (primary)**: Steady descent ep1-47 (58.9%→12.4%), then catastrophic divergence ep50-91 (grad_norm_mean reached **3.2×10^10** at ep91, all 394 batches clipped every epoch). Epochs 92-99 missing from logs (likely NaN). Miraculous recovery at ep100 (8.09%) when cosine LR returned to near-zero.
- **GPU 5 (lr=4e-4)**: Diverged at ep40, grad_norm=**inf** by ep80, **never recovered** — model stuck at 93.5% until timeout. Lower LR delayed onset but made recovery impossible.
- **GPUs 3, 4, 6, 7**: All experienced similar divergence patterns with varying severity. Only runs with enough remaining LR budget post-divergence recovered.

#### Isolation analysis

- **WD effect:** GPU 7 (gc only, no WD) = 8.11% ≈ GPU 2 (gc+WD) = 8.09% → WD has **negligible isolated effect** on DrivAerML
- **T_max effect:** GPU 3 (T_max=20) = 8.16% ≈ GPU 2 (T_max=30) = 8.09% → No significant difference
- **LR effect:** GPU 6 (lr=6e-4) = **7.88%** (best) > GPU 2 (lr=5e-4) = 8.09% > GPU 5 (lr=4e-4) = 16.44% (dead). Higher LR slightly better, lower LR catastrophic.

---

## What happened

**The AirfRANS regularization recipe does NOT transfer to DrivAerML.** The 3D automotive mesh problem has fundamentally different gradient dynamics:

- AirfRANS: gc=1.0 clips ~97% of batches, grad_norm_mean ~10-22 → effective stabilization
- DrivAerML: gc=1.0 clips ALL batches during divergence, grad_norm_mean up to **3.2×10^10** → clipping is powerless

The core issue is scale: DrivAerML's per-batch gradient norms are 5-9 orders of magnitude larger than AirfRANS during instability episodes. The gc=1.0 clip truncates individual batch gradients but cannot prevent the cumulative effect of hundreds of consecutive maximally-clipped updates pushing the model into catastrophic territory.

**However, the AirfRANS result is significant:** 0.001031 is a new best (-16.6%), driven purely by extended training time (360 vs 180 min). The 2L/256d+gc+WD recipe continues to be the optimal AirfRANS config.

**The TandemFoil result is inconclusive** due to I/O contention — with 20 fewer epochs than the baseline, the recipe can't be fairly compared.

## Suggested follow-ups

1. **Rerun DrivAerML primary solo** (no I/O contention) to get a fair epoch-count comparison with the 267-epoch baseline
2. **Try much tighter clipping for DrivAerML** — gc=0.1 or gc=0.01 to actually constrain the 3D mesh gradients
3. **Adaptive gradient clipping** for DrivAerML (e.g., AGC from NFNet paper) — clips based on parameter norm, not a fixed threshold
4. **AirfRANS: lock in 0.001031** — merge this new best as the updated baseline
5. **TandemFoil: rerun solo** with 180-min budget for fair comparison

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2878 because ADVISOR: Closing — dead end. gc=1.0+WD=1e-2 at lr=5e-4 T_max=30 achieved 7.884% (+71% vs 4.619% baseline). DrivAerML gradient norms spike to 3.2×10^10 — gc=1.0 is too tight and blocks learning signal entirely. The T_max=30 baseline without any clipping is strictly better..
